### PR TITLE
Trigger TSV loading immediately after workflow

### LINKED ISSUE #285: Trigger TSV loader as a subDAG or DAG run
## Current Situation

Currently the TSV loader script runs every minute. For a vast majority of its runs (at least at the moment), it skips since there are no files available. Out of the 35,300+ runs we have so far, only 8 have ever run the entire workflow.

## Suggested Improvement

We should instead trigger the TSV loading as part of the provider DAGs. This can be done with either the `TriggerDagRunOperator`[^1] or via SubDAGs[^2].

[^1]: https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/operators/trigger_dagrun/index.html
[^2]: https://airflow.apache.org/docs/apache-airflow/stable/concepts/dags.html#subdags

## Benefit

Both of these strategies would reduce the number of runs this DAG completes significantly. This would make it easier to navigate runs in the UI, track failures, etc.

## Additional context

We should also probably clear out the table of skipped runs at some point. They don't really provide us with much useful information.

## Implementation

- [ ] 🙋 I would be interested in implementing this feature.

### PR DESCRIPTION
**Pull request by AetherUnbound**
## Fixes

Fixes #285 by @AetherUnbound

## Description

At long last! This PR makes some significant changes to the TSV loading process.

**TL;DR**: The TSV loading steps are now connected within the provider workflow DAG and get triggered immediately after the provider finishes/fails.

**The long-winded description**:

The current TSV workflow is:
1. Provider workflow runs
1. TSV loader runs asynchronously every minute and ingests the oldest file

It is more desirable to have synchronous loading for the case where we're ingesting a resource daily/monthly for a few reasons:
- Running the TSV loader every minute results in a very large number of skipped runs (currently), see #309.
- If a problem occurs with a sheet, that sheet is moved to a different location and we have to hunt down the errant file. This also prevents the file from being easily re-run once it's corrected as it must be moved back into the staging location.
- It makes it easier to determine which provider TSV caused the load to fail, since the steps are attached directly to the DAG.

The new change adds one or several [task groups](https://www.astronomer.io/guides/task-groups) on the end of our **basic provider workflow DAG**. I have intentionally left all pieces of the current TSV loader functioning for when we get to enabling the partitioned re-ingestion DAGs. As such, this PR only makes minor tweaks to existing functions; where differing functionality is needed, new functions have been added to support this work.

It's important to note that each provider workflow DAG must now report the media types it supports. For most of our DAGs this is `image`, but any provider script that produces media records beyond that must do so explicitly (see: audio DAGs). The setup is complicated _a bit_ by the fact that a provider script can produce multiple media types (see: Wikimedia commons). In order to handle this case, each media type generates its own ingestion task group. This also helps ensure that the loads are independent of one another.

As described above, the current method ingests the least recently modified file. This new method gets handed the file to ingest explicitly from upstream tasks via [XComs](https://airflow.apache.org/docs/apache-airflow/stable/concepts/xcoms.html). Due to the way our provider API scripts are set up, the location on disk must be retrieved from the provider API module itself. I've added a wrapper function in the DAG which accomplishes this. Ideally with the work from #229 and other efforts, we'll be able to construct the output path using the DAG's execution date and pass that into the `MediaStore` (rather than retrieving it from the store).

In addition to new tests, I've also updated some of the tests to factor out common fixtures.

Phew, okay, I think that's it. There's a lot going on in this PR, even though it isn't a _ton_ of code. Let me know if there's anything I should explain more thoroughly. Much of this is intermediate work to 1) help us get up and running now, 2) pave the way for future work, and 3) prevent this PR from being enormous!

### Screenshots

**Simple, single-media DAG**
Screenshot_2022-02-10_11-32-54

**Multi-media DAG**
Screenshot_2022-02-10_11-33-08

## Suggested review order
1. `dags/common/storage/media.py` & `dags/common/loader/paths.py`- making some attributes public so they can be used in the DAG
2. `dags/common/loader/*` - new, more generic functions for the TSV loading steps
3. `dags/providers/*_workflow.py` - added media types where necessary
4. `dags/common/dag_factory.py` - bulk of the changes, adding on the TSV loading steps to the end of the ingestion workflow
5. `tests/*` and everything else

## Testing Instructions

1. `just test`
1. Run a provider workflow
1. Either "mark successful" or "mark failed" in the UI after some lines have been written
1. Observe that the loader runs immediately after the workflow step and ingests data directly into the database

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Review by obulat — COMMENTED** [review]
The code changes are really nice!

I tried to run jamendo, science_museum, wikimedia_commons and wordpress workflows. For Wikipedia, I get this error: 
```
[2022-02-11, 18:31:16 +03] {taskinstance.py:1700} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/usr/local/airflow/.local/lib/python3.9/site-packages/airflow/models/taskinstance.py", line 1329, in _run_raw_task
    self._execute_task_with_callbacks(context)
  File "/usr/local/airflow/.local/lib/python3.9/site-packages/airflow/models/taskinstance.py", line 1455, in _execute_task_with_callbacks
    result = self._execute_task(context, self.task)
  File "/usr/local/airflow/.local/lib/python3.9/site-packages/airflow/models/taskinstance.py", line 1506, in _execute_task
    result = execute_callable(context=context)
  File "/usr/local/airflow/.local/lib/python3.9/site-packages/airflow/operators/python.py", line 172, in execute
    self.op_kwargs = determine_kwargs(self.python_callable, self.op_args, context)
  File "/usr/local/airflow/.local/lib/python3.9/site-packages/airflow/utils/operator_helpers.py", line 111, in determine_kwargs
    raise ValueError(f"The key {name} in args is part of kwargs and therefore reserved.")
ValueError: The key func in args is part of kwargs and therefore reserved.
```

**Comment by AetherUnbound** [comment]
Oh shoot, thanks - I'll look into that!

**Comment by AetherUnbound** [comment]
Update: this issue is _really_ difficult to track down O_o it has to do with some magic Airflow is doing to manage context.

**Comment by AetherUnbound** [comment]
Figured it out! We had to explicitly pass in `args` as a `kwarg` along with the others in order to prevent Airflow from freaking out. I also addressed a few other issues:
* The loading now skips if no file was written, which is what we discussed in #285 anyway
* The loading table `identifier` was constructed in a way that it could cause a race condition if two providers of the same media type were ran on the same day. Since we're going to be doing that A LOT (ideally), I changed this so the `identifier` now includes the provider name for better distinction.

File: tests/dags/common/test_dag_factory.py [hunk 1]
```diff
@@ -1,8 +1,63 @@
...
+import requests
+from airflow.models import TaskInstance
+from common import dag_factory
+
+from tests.dags.common.test_resources import fake_provider_module
+
+
+@pytest.mark.parametrize(
+    "func, media_types, stores",
+    [
```
Comment (krysal) [inline tests/dags/common/test_dag_factory.py:13]: Great testing, covering a wide range of paths!

File: openverse_catalog/dags/database/loader_workflow.py [hunk 2]
```diff
@@ -104,12 +104,19 @@
         """
         ),
     )
+
+    media_type_xcom = (
+        "{{ ti.xcom_pull(task_ids='%s', key='media_type') }}"
+        % stage_oldest_tsv_file.task_id
```
Comment (sarayourfriend) [inline openverse_catalog/dags/database/loader_workflow.py:110]: I always preferred a good old `.format` call than the `%` version when possible. Too many rules to remember with the `%` method.

```suggestion
        "{{{{ ti.xcom_pull(task_ids='{task_id}', key='media_type') }}}}".format(
            task_id=stage_oldest_tsv_file.task_id
        )
```

Or even just an `f` string:

```suggestion
        "{{{{ ti.xcom_pull(task_ids='{stage_oldest_tsv_file.task_id}', key='media_type') }}}}"
```

I suppose some folks don't like the curly brace escaping though?

**Review by krysal — APPROVED** [review]
Tested with StockSnap and Jamendo scripts and it works perfectly! Thanks for the detailed explanation and comments 🌟

File: openverse_catalog/dags/common/dag_factory.py [hunk 3]
```diff
@@ -71,8 +126,15 @@ def create_provider_api_workflow(
                       be run (if `dated=True`).
     dagrun_timeout:   datetime.timedelta giving the total amount of time
                       a given dagrun may take.
+    doc_md:           string which should be used for the DAG's documentation markdown
+    media_types:      list describing the media type(s) that this provider handles
+                      (e.g. ("image",), ["audio"], ["image", "audio"], etc.)
```
Comment (sarayourfriend) [inline openverse_catalog/dags/common/dag_factory.py:131]: Mixing the types in the documentation is confusing to me for whatever reason... or at least it took a little longer of reading it to understand what was going on. Maybe some back ticks would help?

```suggestion
                      (e.g. `("image",)`, `["audio"]`, `["image", "audio"]`, etc.)
```

It already being nested inside of grammatical parenthesis makes the tuple version not stand out as much.

File: openverse_catalog/dags/common/dag_factory.py [hunk 4]
```diff
@@ -87,20 +149,71 @@ def create_provider_api_workflow(
...
+                    python_callable=sql.drop_load_table,
+                    op_kwargs={
+                        "postgres_conn_id": DB_CONN_ID,
+                        "identifier": identifier,
+                        "media_type": media_type,
+                    },
+                    trigger_rule=TriggerRule.ALL_DONE,
+                )
+                [create_loading_table, copy_to_s3] >> load_from_s3
+                [create_loading_table, load_from_s3] >> drop_loading_table
```
Comment (sarayourfriend) [inline openverse_catalog/dags/common/dag_factory.py:214]: What's going on here? :magic_wand:

File: openverse_catalog/dags/common/dag_factory.py [hunk 5]
```diff
@@ -22,19 +32,64 @@
...
+    # must first pull the stores from the module.
+    for media_type in media_types:
+        if not (store := getattr(module, f"{media_type}_store", None)):
+            continue
+        stores[media_type] = store
+
+    if len(stores) != len(media_types):
+        raise ValueError(
+            f"Expected a store of the following types in {module.__name__} "
+            f"but none were found: {media_types}"
```
Comment (stacimc) [inline openverse_catalog/dags/common/dag_factory.py:68]: It might be useful to log which stores were missing, or else change the wording slightly for clarity (`none were found` -> `not all were found`?).

File: openverse_catalog/dags/common/dag_factory.py (hunk 4 above)
Comment (stacimc) [reply openverse_catalog/dags/common/dag_factory.py:214]: I think this is [setting the dependencies](https://airflow.apache.org/docs/apache-airflow/stable/concepts/tasks.html#relationships) for the tasks within the `load_data` TaskGroup. 

@AetherUnbound I'm curious if this is logically equivalent to:
```
[create_loading_table, copy_to_s3] >> load_from_s3 >> drop_loading_table
```
? I think it's still clear this way just wondering what the difference is.

I understand the sequence here but it took a struggle, mostly because of unfamiliarity with Airflow (and the actual process!). I think it would be worth adding some comments (or maybe expanding the docstring for the factory method) to describe the flow.

File: openverse_catalog/dags/common/loader/s3.py [hunk 6]
```diff
@@ -28,6 +38,34 @@ def copy_file_to_s3_staging(
...
+    ti,
+):
+    """
+    Copy a TSV file to S3 with the given prefix.
+    The TSV's version is pushed to the `tsv_version` XCom, and the constructed
+    S3 key is pushed to the `s3_key` XCom.
+    The TSV is removed after the upload is complete.
+    """
+    if tsv_file_path is None:
+        raise FileNotFoundError(f"Expected file {tsv_file_path} was not provided")
```
Comment (stacimc) [inline openverse_catalog/dags/common/loader/s3.py:55]: In this case it's not so much that the file wasn't provided but the path itself, so there shouldn't be a `tsv_file_path` to log:
```suggestion
        raise FileNotFoundError(f"No TSV file path was provided.")
```

File: openverse_catalog/dags/database/loader_workflow.py (hunk 2 above)
Comment (obulat) [reply openverse_catalog/dags/database/loader_workflow.py:110]: I would prefer fstrings if they work here.

File: openverse_catalog/dags/common/dag_factory.py [hunk 7]
```diff
@@ -22,19 +32,64 @@
...
+    main_function: Callable,
+    default_args: Optional[Dict] = None,
+    start_date: datetime = datetime(1970, 1, 1),
+    max_active_tasks: int = 1,
+    schedule_string: str = "@daily",
+    dated: bool = True,
+    day_shift: int = 0,
+    dagrun_timeout: timedelta = timedelta(hours=12),
+    doc_md: str = "",
+    media_types: List[str] = ("image",),
```
Comment (obulat) [inline openverse_catalog/dags/common/dag_factory.py:92]: Does `List` type also include tuples? I understand that the default value cannot be a list because it's mutable. Does the type declaration match the default value?

**Review by obulat — COMMENTED** [review]
What happens when the `pull_data` task is manually set to successful? Does it not run `commit` for any media that hadn't been written to disk yet?
Because when I run `wikimedia` DAG and set `pull_mixed_data` to `successful`, `load_image_data` tasks are all green, but the `load_audio_data` `copy_to_s3` and `load_from_s3` ares skipped. I guess it's because there is not file written to the disk? It would be nice to skip `create_loading_table` task if there is no data available, if that's at all possible as it is run concurrently with `copy_to_s3`...

File: openverse_catalog/dags/common/dag_factory.py [hunk 8]
```diff
@@ -22,19 +32,64 @@
...
-    dated=True,
-    day_shift=0,
-    dagrun_timeout=timedelta(hours=12),
-    doc_md="",
+    dag_id: str,
+    main_function: Callable,
+    default_args: Optional[Dict] = None,
+    start_date: datetime = datetime(1970, 1, 1),
+    max_active_tasks: int = 1,
+    schedule_string: str = "@daily",
```
[the remark this answers is not in the document]
Comment (AetherUnbound) [reply openverse_catalog/dags/common/dag_factory.py:87]: It does not, unfortunately. Literals could be helpful here, but the `@daily`/`@monthly` schedule intervals are actually just [aliases to cron strings, which is what this field represents](https://airflow.apache.org/docs/apache-airflow/stable/dag-run.html?highlight=schedule%20interval#dag-runs).

File: openverse_catalog/dags/common/dag_factory.py (hunk 4 above)
Comment (sarayourfriend) [reply openverse_catalog/dags/common/dag_factory.py:214]: :exploding_head: 

This is cryptic, but I understand that if you know Airflow already then it's pretty normal syntax! I'm happy to just accept it as is (and learn more about it at a later date!) :slightly_smiling_face:

File: openverse_catalog/dags/common/loader/loader.py [hunk 9]
```diff
@@ -31,3 +31,19 @@ def load_s3_data(
...
+    media_type,
+    tsv_version,
+    identifier,
+):
+    sql.load_s3_data_to_intermediate_table(
+        postgres_conn_id, bucket, key, identifier, media_type
+    )
+    sql.upsert_records_to_db_table(
+        postgres_conn_id, identifier, media_type=media_type, tsv_version=tsv_version
+    )
```
[the remark this answers is not in the document]
Comment (AetherUnbound) [reply openverse_catalog/dags/common/loader/loader.py:49]: I'm not sure the original context/rationale for this decision, but an intermediate table has a few benefits:
1. Isolation from the primary table, so it isn't impacted by the additional time it might take to get the data from S3, only the time it takes to upsert the records from another table
2. We may have loads that affect a large enough number of records that couldn't adequately fit in memory and would be easier to aggregate on disk
3. The intermediate tables get indexed, and those indices can be carried forward into the primary table & reduce the upsertion time

I'm not sure we're at the place _yet_ where we're processing that much data, but we haven't run all the providers regularly so it's possible!

[the remark this answers is not in the document]
Comment (sarayourfriend) [reply openverse_catalog/dags/common/loader/loader.py:49]: Okay, cool! Thanks for the explanation of benefits :grinning:

File: openverse_catalog/dags/common/loader/s3.py [hunk 10]
```diff
@@ -28,6 +38,34 @@ def copy_file_to_s3_staging(
...
+    tsv_file = Path(tsv_file_path)
+    if not tsv_file.exists():
+        raise AirflowSkipException(f"TSV file {tsv_file} had no data.")
```
[the remark this answers is not in the document]
Comment (AetherUnbound) [reply openverse_catalog/dags/common/loader/s3.py:58]: Yes! That's a more appropriate message. This case typically occurs when there is no data to process, but the current message is misleading.

File: openverse_catalog/dags/database/loader_workflow.py (hunk 2 above)
Comment (AetherUnbound) [reply openverse_catalog/dags/database/loader_workflow.py:110]: I was trying to avoid having to escape the brackets, since it ends up with a lot of `{{{{` :sweat_smile: If y'all prefer it though, I'll change it!

File: openverse_catalog/dags/common/dag_factory.py (hunk 3 above)
Comment (AetherUnbound) [reply openverse_catalog/dags/common/dag_factory.py:131]: Ah, totally fair. My intent here was to point out that any sequence type is accepted. We're using a tuple as the default argument because lists are mutable and mutable default arguments are probably the most common "gotcha" in Python[^1]. If you think that this example is better served by consistent types, I'm happy to change it!

[^1]: https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

File: openverse_catalog/dags/common/loader/s3.py (hunk 6 above)
Comment (AetherUnbound) [reply openverse_catalog/dags/common/loader/s3.py:55]: :woman_facepalming: totally true, I'll update this!

File: openverse_catalog/dags/common/dag_factory.py (hunk 7 above)
Comment (AetherUnbound) [reply openverse_catalog/dags/common/dag_factory.py:92]: Ugh, even after [all my best intentions](https://github.com/WordPress/openverse-catalog/pull/357#discussion_r809458638) I still get this type wrong! I'll fix this to make it more generic.

**Comment by AetherUnbound** [comment]
> What happens when the `pull_data` task is manually set to successful? Does it not run `commit` for any media that hadn't been written to disk yet? Because when I run `wikimedia` DAG and set `pull_mixed_data` to `successful`, `load_image_data` tasks are all green, but the `load_audio_data` `copy_to_s3` and `load_from_s3` ares skipped. I guess it's because there is not file written to the disk? It would be nice to skip `create_loading_table` task if there is no data available, if that's at all possible as it is run concurrently with `copy_to_s3`...

Ha, I know exactly the case you're talking about! The wikimedia dag is `dated`, so IIRC it's just pulling _all_ data for that day and sorting it into image & audio bins depending on the media type. I think it's likely that the provider script just hadn't come across any audio data yet, so there was nothing to write to disk.

On your second question: I think that'd be possible, we'd need some additional logic to determine whether the file has any data. That could be done as part of the create table step, or perhaps it'd be better served as its own step which skips the rest of the workflow. As it stands currently, it's easier to unconditionally create the table & delete it at the end of the load, even if nothing happens with it.

File: openverse_catalog/dags/common/dag_factory.py (hunk 4 above)
Comment (stacimc) [reply openverse_catalog/dags/common/dag_factory.py:214]: Oh that's perfect! Definitely, I think copying it over would make a lot of sense 😄

File: openverse_catalog/dags/common/dag_factory.py (hunk 5 above)
Comment (AetherUnbound) [reply openverse_catalog/dags/common/dag_factory.py:68]: How is this?
image

File: openverse_catalog/dags/common/loader/s3.py [hunk 11]
```diff
@@ -28,6 +35,34 @@ def copy_file_to_s3_staging(
...
+    if tsv_file_path is None:
+        raise FileNotFoundError("No TSV file path was provided")
+    tsv_file = Path(tsv_file_path)
+    if not tsv_file.exists():
+        raise AirflowSkipException(f"TSV file {tsv_file} does not exist.")
+    tsv_version = paths.get_tsv_version(tsv_file_path)
+    s3_key = f"{s3_prefix}/{tsv_file.name}"
+    logger.info(f"Uploading {tsv_file_path} to {s3_bucket}:{s3_key}")
+    s3 = S3Hook(aws_conn_id=aws_conn_id)
+    s3.load_file(tsv_file_path, s3_key, bucket_name=s3_bucket)
```
Comment (rbadillap) [inline openverse_catalog/dags/common/loader/s3.py:60]: Is it possible that there can be a file with the same name already placed in the bucket? I'm wondering if worth the chance to run some validation to make sure if there's a file with the same name.

[As per the source code](https://github.com/apache/airflow/blob/4dd751d2fc3f091417146079d7d1b3ac6dd86c9c/airflow/providers/amazon/aws/hooks/s3.py#L497) the default behavior is to overwrite the file, is this the expected behavior?

Comment (AetherUnbound) [reply openverse_catalog/dags/common/loader/s3.py:60]: Good question! The S3 key is constructed using the provider, media type, TSV version, and timestamp for the run:
https://github.com/WordPress/openverse-catalog/blob/df7f68460a7ceb14cc39a02743ba8190411c017c/openverse_catalog/dags/common/storage/media.py#L175

The only case where we'll have a duplicate file is if the TSV is attempted to be uploaded after it has already been successfully uploaded. This would require some pretty intentional manual intervention to accomplish, so I think we're good :slightly_smiling_face: 

Also it looks like the default is actually to raise an error if the file exists, so we won't perform any unintentional overwrites either:
https://github.com/apache/airflow/blob/4dd751d2fc3f091417146079d7d1b3ac6dd86c9c/airflow/providers/amazon/aws/hooks/s3.py#L483

File: openverse_catalog/dags/common/dag_factory.py (hunk 3 above)
Comment (sarayourfriend) [reply openverse_catalog/dags/common/dag_factory.py]: I do think consistent types would be good as well as the back-ticks to help delineate the code parens from the grammatical ones. If anything, always using a tuple at least communicates to the reader that the list will not be modified.

**Comment by obulat** [comment]
> Ha, I know exactly the case you're talking about! The wikimedia dag is dated, so IIRC it's just pulling all data for that day and sorting it into image & audio bins depending on the media type. I think it's likely that the provider script just hadn't come across any audio data yet, so there was nothing to write to disk.

Yes, I suspected that it hadn't gotten any audio, so I reran the script untill I saw the `Got x audio so far` logging message, with x being around 30. Then I set the `pull_mixed_media` task to sucessful in the UI

**Review by obulat — APPROVED** [review]

**Comment by AetherUnbound** [comment]
> Yes, I suspected that it hadn't gotten any audio, so I reran the script until I saw the `Got x audio so far` logging message, with x being under 100.

Ah, in that case you're totally write. We only write to buffer every 100 records, so if we never reach 100 records, the only write happens at the end unfortunately.
 
> I wonder what happens in case of an error during the pull stage? All the data would be lost, and hopefully recollected in a retry, right? I wish there was a way to re-start a failed pull task from where it failed, and not from the start, but it probably goes against Airflow philosophy of "tasks should always be deterministic and return the same result for the same inputs".

That's a good thing to be thinking about! It's possible that we could store some recovery information on failure (maybe the current set of query params?) and pick those up if the task is retried. Currently, if the data pull fails, one of two things will happen:
1. If =>100 records have been processed, the generated TSV will exist on disk with the last buffer save, and that will be picked up by the TSV loader.
2. If <100 records have been processed, no data has been written so no data is written to disk (I believe). 

I think it's worth looking into how the latter can be changed! I'll make an issue for it.

**Review by stacimc — APPROVED** [review]
Looks great 🚀 

> And frankly, I guess it's OK because normally you wouldn't stop the task manually.
We really _do_ seem to get bit by this a lot! Thanks for all the detailed explanations in this discussion 😄
